File: test/e2e/Apps/BasicDotNetIsolated/RewindOrchestration.cs
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System.Collections.Concurrent;
using System.Net;
using Microsoft.Azure.Functions.Worker;
using Microsoft.Azure.Functions.Worker.Http;
using Microsoft.DurableTask;
using Microsoft.DurableTask.Client;
using Microsoft.DurableTask.Entities;
using Microsoft.Extensions.Logging;

namespace Microsoft.Azure.Durable.Tests.E2E;

public static class RewindOrchestration
{
    private static readonly ConcurrentDictionary<string, int> invocationCounts = [];
    private static readonly EntityInstanceId entityId = new(nameof(InvocationCounterEntity), "entity");

    [Function(nameof(RewindParentOrchestration))]
    public static async Task<ConcurrentDictionary<string, int>> RewindParentOrchestration(
        [OrchestrationTrigger] TaskOrchestrationContext context)
    {
        OrchestrationInput? input = context.GetInput<OrchestrationInput>();
        if (input?.Name == "run")
        {
            await context.CreateTimer(context.CurrentUtcDateTime.AddMinutes(10), CancellationToken.None);
            return [];
        }
        else if (input?.Name == "complete")
        {
            return [];
        }
        else if (input?.Name == "fail")
        {
            Task[] subOrchestrationTasks = 
            {
                context.CallSubOrchestratorAsync<string>(
                    nameof(SucceedSubOrchestration), "succeed_sub_1"),
                context.CallSubOrchestratorAsync<string>(
                    nameof(FailParentSubOrchestration), new OrchestrationInput("fail_parent_sub_1", input.NumFailures, input.CallEntities)),
                context.CallSubOrchestratorAsync<string>(
                    nameof(FailParentSubOrchestration), new OrchestrationInput("fail_parent_sub_2", input.NumFailures, input.CallEntities)),
                context.CallSubOrchestratorAsync<string>(
                    nameof(SucceedSubOrchestration), "succeed_sub_2")
            };
            await Task.WhenAll(subOrchestrationTasks);
            return invocationCounts;
        }
        else
        {
            throw new ArgumentException("Invalid input");
        }
    }

    [Function(nameof(FailChildSubOrchestration))]
    public static async Task<string> FailChildSubOrchestration(
        [OrchestrationTrigger] TaskOrchestrationContext context)
    {
        OrchestrationInput input = context.GetInput<OrchestrationInput>()!;
        List<Task> tasks =
        [
            context.CallActivityAsync<string>(nameof(SucceedActivity), input.Name + "_succeed_activity"),
            context.CallActivityAsync<string>(nameof(FailActivity), new OrchestrationInput(input.Name + "_fail_activity_1", input.NumFailures, input.CallEntities)),
            context.CallActivityAsync<string>(nameof(FailActivity), new OrchestrationInput(input.Name + "_fail_activity_2", input.NumFailures, input.CallEntities))
        ];
        if (input.CallEntities)
        {
            tasks.Add(context.Entities.SignalEntityAsync(entityId, input.Name + "_signal_entity"));
            tasks.Add(context.Entities.CallEntityAsync(entityId, input.Name + "_call_entity"));
        }
        await Task.WhenAll(tasks);
        return "Ok, sub done!";
    }

    [Function(nameof(FailParentSubOrchestration))]
    public static async Task<string> FailParentSubOrchestration(
        [OrchestrationTrigger] TaskOrchestrationContext context)
    {
        OrchestrationInput input = context.GetInput<OrchestrationInput>()!;
        await context.CallActivityAsync<string>(nameof(SucceedActivity), input.Name + "_succeed_activity");
        if (input.CallEntities)
        {
            await context.Entities.CallEntityAsync(entityId, input.Name + "_call_entity");
        }
        await context.CallSubOrchestratorAsync<string>(nameof(FailChildSubOrchestration), new OrchestrationInput(input.Name + "_child", input.NumFailures, input.CallEntities));
        return "Ok, sub done!";
    }

    [Function(nameof(SucceedSubOrchestration))]
    public static async Task<string> SucceedSubOrchestration(
        [OrchestrationTrigger] TaskOrchestrationContext context)
    {
        await context.CallActivityAsync<string>(nameof(SucceedActivity), context.GetInput<string>() + "_succeed_activity");
        return "Ok, sub done!";
    }

    [Function(nameof(SucceedActivity))]
    public static string SucceedActivity([ActivityTrigger] string input, FunctionContext executionContext)
    {
        UpdateInvocationCount(input);
        return $"Hello, {input}!";
    }

    [Function(nameof(FailActivity))]
    public static string FailActivity([ActivityTrigger] OrchestrationInput failInfo, FunctionContext executionContext)
    {
        if (UpdateInvocationCount(failInfo.Name!) <= failInfo.NumFailures)
        {
            throw new Exception("Failure!");
        }
        return "Success!";
    }

    [Function(nameof(RewindInstance))]
    public static async Task<HttpResponseData> RewindInstance(
        [HttpTrigger(AuthorizationLevel.Anonymous, "get", "post")] HttpRequestData req,
        [DurableClient] DurableTaskClient client,
        string instanceId)
    {
        string rewindReason = "Rewinding the instance for testing.";
        try
        {
            await client.RewindInstanceAsync(instanceId, rewindReason);

        }
        catch (InvalidOperationException)
        {
            return req.CreateResponse(HttpStatusCode.PreconditionFailed);
        }
        catch (ArgumentException)
        {
            return req.CreateResponse(HttpStatusCode.NotFound);
        }
        catch (NotImplementedException)
        {
            return req.CreateResponse(HttpStatusCode.NotImplemented);
        }
        return req.CreateResponse(HttpStatusCode.OK);
    }

    [Function(nameof(HttpStart_RewindOrchestration))]
        public static async Task<HttpResponseData> HttpStart_RewindOrchestration(
            [HttpTrigger(AuthorizationLevel.Anonymous, "get", "post")] HttpRequestData req,
            [DurableClient] DurableTaskClient client,
            FunctionContext executionContext,
            string input,
            int numFailures,
            bool callEntities,
            bool? delay)
    {
        invocationCounts.Clear();
        ILogger logger = executionContext.GetLogger(nameof(HttpStart_RewindOrchestration));

        string instanceId = await client.ScheduleNewOrchestrationInstanceAsync(
            nameof(RewindParentOrchestration),
            new OrchestrationInput(input, numFailures, callEntities),
            delay == true ? new StartOrchestrationOptions { StartAt = DateTimeOffset.UtcNow.AddMinutes(1) } : null); 

        logger.LogInformation("Started orchestration with ID = '{instanceId}'.", instanceId);

        // Returns an HTTP 202 response with an instance management payload.
        // See https://learn.microsoft.com/azure/azure-functions/durable/durable-functions-http-api#start-orchestration
        return await client.CreateCheckStatusResponseAsync(req, instanceId);
    }

    [Function(nameof(InvocationCounterEntity))]
    public static Task InvocationCounterEntity([EntityTrigger] TaskEntityDispatcher dispatcher)
    {
        return dispatcher.DispatchAsync(operation =>
        {
            UpdateInvocationCount(operation.Name);
            return default;
        });
    }

    private static int UpdateInvocationCount(string key)
    {
        if (!invocationCounts.TryGetValue(key, out int invocationCount))
        {
            invocationCount = 0;
            invocationCounts[key] = invocationCount;
        }
        invocationCounts[key] = ++invocationCount;
        return invocationCount;
    }

    public class OrchestrationInput(string name, int numFailures, bool callEntities)
    {
        public string? Name { get; set; } = name;

        public int NumFailures { get; set; } = numFailures;

        public bool CallEntities { get; set; } = callEntities;
    }
}

File: test/e2e/Apps/BasicNode/src/functions/RewindOrchestration.ts
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { app, HttpHandler, HttpRequest, HttpResponse, InvocationContext } from '@azure/functions';
import * as df from 'durable-functions';
import { ActivityHandler, OrchestrationContext, OrchestrationHandler, EntityContext, EntityHandler } from 'durable-functions';
 
const invocationCounts: Map<string, number> = new Map();
const entityId = new df.EntityId('InvocationCounterEntity', 'entity');

// Orchestration
const RewindParentOrchestration: OrchestrationHandler = function* (context: OrchestrationContext) {
    const input = context.df.getInput<OrchestrationInput>();

    if (input.name == 'run')
    {
        // Create a timer for 10 minutes
        yield context.df.createTimer(new Date(Date.now() + 60 * 10 * 1000));
    }
    else if (input.name == 'complete')
    {
        return {};
    }
    else if (input.name == 'fail')
    {
        const subOrchestrationTasks = [
            context.df.callSubOrchestrator('SucceedSubOrchestration', 'succeed_sub_1'),
            context.df.callSubOrchestrator(
                'FailParentSubOrchestration',
                new OrchestrationInput('fail_parent_sub_1', input.numFailures, input.callEntities)),
            context.df.callSubOrchestrator(
                'FailParentSubOrchestration',
                new OrchestrationInput('fail_parent_sub_2', input.numFailures, input.callEntities)),
            context.df.callSubOrchestrator('SucceedSubOrchestration', 'succeed_sub_2')
        ];

        yield context.df.Task.all(subOrchestrationTasks);
        return invocationCounts;
    }
    else
    {
        throw new Error('Invalid input');
    }
};
df.app.orchestration('RewindParentOrchestration', RewindParentOrchestration);

// Suborchestrations
const FailParentSubOrchestration: OrchestrationHandler = function* (context: OrchestrationContext) {
    const input = context.df.getInput<OrchestrationInput>();
    yield context.df.callActivity('SucceedActivity', input.name + '_succeed_activity');
    if (input.callEntities)
    {
        yield context.df.callEntity(entityId, input.name + '_call_entity');
    }
    yield context.df.callSubOrchestrator(
        'FailChildSubOrchestration',
        new OrchestrationInput(input.name + '_child', input.numFailures, input.callEntities));
}
df.app.orchestration('FailParentSubOrchestration', FailParentSubOrchestration);

const FailChildSubOrchestration: OrchestrationHandler = function* (context: OrchestrationContext) {
    const input = context.df.getInput<OrchestrationInput>();

    if (input.callEntities)
    {
        context.df.signalEntity(entityId, input.name + '_signal_entity');
    }

    const activityAndEntityTasks  = [
        context.df.callActivity('SucceedActivity', input.name + '_succeed_activity'),
        context.df.callActivity('FailActivity', new OrchestrationInput(input.name + '_fail_activity_1', input.numFailures, input.callEntities)),
        context.df.callActivity('FailActivity', new OrchestrationInput(input.name + '_fail_activity_2', input.numFailures, input.callEntities)),
    ];

    if (input.callEntities)
    {
        activityAndEntityTasks.push(context.df.callEntity(entityId, input.name + '_call_entity'));
    }

    yield context.df.Task.all(activityAndEntityTasks);
}
df.app.orchestration('FailChildSubOrchestration', FailChildSubOrchestration);

const SucceedSubOrchestration: OrchestrationHandler = function* (context: OrchestrationContext) {
    yield context.df.callActivity('SucceedActivity', context.df.getInput<string>() + '_succeed_activity');
}
df.app.orchestration('SucceedSubOrchestration', SucceedSubOrchestration);

// Activities
const SucceedActivity: ActivityHandler = (input: string): string => {
    UpdateInvocationCount(input);
    return 'Hello ' + input;
};
df.app.activity('SucceedActivity', { handler: SucceedActivity });

const FailActivity: ActivityHandler = (input: OrchestrationInput): string => {
    if (UpdateInvocationCount(input.name) <= input.numFailures)
    {
        throw new Error('Failure!');
    }
    return 'Success!';
};
df.app.activity('FailActivity', { handler: FailActivity });

// HTTP Rewind Instance
const RewindInstance: HttpHandler = async (req: HttpRequest, context: InvocationContext): Promise<HttpResponse> => {
    const client = df.getClient(context);
    let body = await req.json().catch(() => ({}));
    let instanceId = await body['instanceId'] || req.query.get('instanceId') || req.params['instanceId'];
    const reason = 'Rewinding the instance for testing.';
    try {
        await client.rewind(instanceId, reason);
        return new HttpResponse({ status: 200 });
    } catch (ex: any) {
        return new HttpResponse({ body: String(ex), status: 400, headers: { 'Content-Type': 'text/plain' } });
    }
};

app.http('RewindInstance', {
    route: 'RewindInstance',
    methods: ['GET', 'POST'],
    extraInputs: [df.input.durableClient()],
    handler: RewindInstance,
});

const HttpStart_RewindOrchestration: HttpHandler = async (request: HttpRequest, context: InvocationContext): Promise<HttpResponse> => {
    const client = df.getClient(context);
    invocationCounts.clear();

    const instanceId: string = await client.startNew(
        'RewindParentOrchestration',
        { input: new OrchestrationInput(
            request.params.input,
            Number.parseInt(request.params.numFailures),
            request.params.callEntities == 'true'
        )});

    context.log(`Started orchestration with ID = '${instanceId}'.`);

    return client.createCheckStatusResponse(request, instanceId);
};

app.http('HttpStart_RewindOrchestration', {
    route: 'HttpStart_RewindOrchestration',
    extraInputs: [df.input.durableClient()],
    handler: HttpStart_RewindOrchestration,
});

function UpdateInvocationCount(key) {
  let invocationCount = invocationCounts.get(key) || 0; // get existing count or 0
  invocationCounts.set(key, ++invocationCount);
  return invocationCount;
}

// Entity
const InvocationCounterEntity: EntityHandler<{state: string}> = (context: EntityContext<{state: string}>) => {
    UpdateInvocationCount(context.df.operationName);  
};
df.app.entity('InvocationCounterEntity', InvocationCounterEntity);

class OrchestrationInput
{
    name: string;
    numFailures: number;
    callEntities: boolean;

    constructor(name: string, numFailures: number, callEntities: boolean)
    {
        this.name = name;
        this.numFailures = numFailures;
        this.callEntities = callEntities;
    }
}